### ISSUE TITLE
Support image relocation

### ISSUE THREAD

**Issue by glyn** [deislabs/duffle#668]
The externals are still to be defined and will be added here in due course.

The intention is to use https://github.com/google/go-containerregistry for accessing OCI image layouts and remote repositories.

**Comment by silvin-lubecki** [comment]
I think the use case you described is exactly what https://github.com/docker/cnab-to-oci does (@simonferquel please jump in if I'm wrong).
You can pull the bundle from a registry, then push it to another (private or not) registry, and `cnab-to-oci` will update the image references. It even support [image promotion pattern](https://docs.docker.com/ee/dtr/user/promotion-policies/internal-promotion/).

With @simonferquel we planned to integrate this library to `duffle` (it's already integrated in `docker-app`), had no time yet to do it.

**Comment by glyn** [comment]
> I think the use case you described is exactly what https://github.com/docker/cnab-to-oci does (@simonferquel please jump in if I'm wrong).
> You can pull the bundle from a registry, then push it to another (private or not) registry, and `cnab-to-oci` will update the image references. It even support [image promotion pattern](https://docs.docker.com/ee/dtr/user/promotion-policies/internal-promotion/).
> 
> With @simonferquel we planned to integrate this library to `duffle` (it's already integrated in `docker-app`), had no time yet to do it.

We have a use case where the bundle is _not_ stored in a registry. Will that also be supported by cnab-to-oci?

**Comment by silvin-lubecki** [comment]
> We have a use case where the bundle is not stored in a registry. Will that also be supported by cnab-to-oci?

I don't understand the use case described earlier then, it says you want to relocate to another registry.
What do you want to relocate then, if it is not towards another registry?
Or maybe you mean you will relocate the images, but not the bundle itself?
That's not how `cnab-to-oci` works. See it more as pushing the thick bundle (bundle + the images) to a registry, as a whole.

**Comment by glyn** [comment]
> Or maybe you mean you will relocate the images, but not the bundle itself?

Correct! So it _seems_ that `cnab-to-oci` won't support our use case. Agreed?

**Comment by simonferquel** [comment]
Yes, the goal of cnab-to-oci is to be able to publish bundles with guaranteed immutability from the bundle metadata down to the actual component and invocation images.
It also guarantees that if you have pull rights for the bundle, you also have pull rights for the invocation image and component images.

**Comment by silvin-lubecki** [comment]
> We have a use case where the bundle is not stored in a registry. 

@glyn could you describe this use case? 
It's interesting for `cnab-to-oci` as we may have all the building blocks to do it, and just need to add a top-level function to handle this relocation.

Maybe you can play with [FixupBundle](https://godoc.org/github.com/docker/cnab-to-oci/remotes#FixupBundle) (which already relocates the images and patches the bundle itself with the digested references) and tell me if maybe it fits your need?

**Comment by glyn** [comment]
Thanks @simonferquel. I agree that goal is laudable and maybe our use case will change in the future to align with that goal. 

> @glyn could you describe this use case?

@silvin-lubecki We currently ship distributions of our software as a zipped archive plus a CLI program. Users relocate images to a (typically) private registry using the CLI. There are a number of steps involved and the process is a little complex to document and for the user to follow.

In the future, we are considering shipping a CNAB instead, either as a thin bundle for users with a DMZ, from which they can relocate images (but _not_ the CNAB itself -- see below) from public repositories to a (typically) private registry, or as a fat bundle for users with no DMZ, in which case they can relocate from an OCI image layout in the bundle to a (typically) private registry.

We expect the user to use a standard CNAB runtime, such as `duffle`, to perform this relocation as well as installation. (The purpose of this issue is to enable `duffle` to perform this style of relocation.)

We don't currently have a registry in which to host a CNAB and, as CNAB is still being developed, we don't want have the user push the CNAB to their private registry. From a registry perspective, it should just look very similar to today: images are pushed to a private registry and our software pulls the images from the private registry.

**Comment by glyn** [comment]
> Maybe you can play with [FixupBundle](https://godoc.org/github.com/docker/cnab-to-oci/remotes#FixupBundle) (which already relocates the images and patches the bundle itself with the digested references) and tell me if maybe it fits your need?

I took a look at `FixupBundle` and the associated command line. These perform a many-one mapping of repositories, whereas we currently use a one-one mapping and would like `duffle` to do the same (see "Implementation Notes" [above](https://github.com/deislabs/duffle/issues/668#issue-422338289)).

The main downside of a many-one mapping is that the mapped image names are effectively indistinguishable from each other from a human readability point of view and certainly not easy to relate to the corresponding original image names. In our case, when we relocate kubernetes configuration files containing a number of image names, it's important to distinguish between mapped image names and preserve a readable connection to the original image names. This ensures that the relocated configuration files remain readable by users.

Note that the mapping we currently employ ensures the same authentication/authorisation property as mapping to a single repository: the mapped repositories are authenticated with and authorised to a single user.

**Comment by simonferquel** [comment]
@glyn This is definitely a valid issue. One thing that we plan to do (at least for docker-app bundles) is to populate a well known annotation on kubernetes/swarm resources that we create with the original image name. 
In the future, we'll add support for the Docker CLI so that things like `docker stack ps` reports both the real image and the original one.

**Comment by glyn** [comment]
Let's discuss your [comment on PR 671](https://github.com/deislabs/duffle/pull/671#issuecomment-483558841) here rather than in the PR:

> Hello [Glyn], as we [discussed](https://github.com/deislabs/duffle/issues/668#issuecomment-475160071) in the issue, I think we should try to improve [cnab-to-oci Fixup](https://github.com/docker/cnab-to-oci/blob/master/remotes/fixup.go) instead.
> 
> If I understand it correctly, the only blocking point you saw in `cnab-to-oci` is that we mount/copy all the images in the same repository, and in your case you need to push all the images to their "own" repository? My guess is that this can be easily implemented in the current Fixup command, it is just a matter of where we push the images, but all the logic of pushing is already there.
> @simonferquel and I would be pleased to help you write this PR against cnab-to-oci 👍
> 
> A second argument for not merging this PR as-is, is that it adds a dependency on `github.com/google/go-containerregistry`, which I think will "duplicate" with the `containerd` library used to [push/pull bundles to a registry](https://github.com/deislabs/duffle/pull/681). @radu-matei any thoughts on that point?

There are a number of requirements for `duffle relocate` which I’d want to make sure are satisfied by an implementation based on cnab-to-oci:
1. Must preserve content digest of each image during relocation.
2. Must support a one-one repository mapping with similar usability to that provided by the current PR.
3. Must support copying images from OCI image layouts to remote registries so that we can relocate thick bundles.
4. Must be able to relocate a bundle which is _not_ stored in a registry. Isn’t this contrary to the over-arching design of cnab-to-oci?
5. Should have the same authentication configuration as docker CLI.

**Comment by simonferquel** [comment]
> Must preserve content digest of each image during relocation.

This is the case, except when using platform filtering (which is off by default)
> Must support a one-one repository mapping with similar usability to that provided by the current PR.

That is where your PR would come in :)
> Must support copying images from OCI image layouts to remote registries so that we can relocate thick bundles.

Not yet supported, although we want to do that. (we already use containerd under the hood, and it is actually really easy to do
>Must be able to relocate a bundle which is not stored in a registry. Isn’t this contrary to the over-arching design of cnab-to-oci?

For now the implementation tries to implement an end-to-end whole bundle relocation. However, we'd like for your PR to refactor the code to bring a little bit of flexibility here (while still providing easy push/relocation APIs for bundle-level operations)
> Should have the same authentication configuration as docker CLI.

Already the case

I am convinced the amount of work for bringing this flexibility to cnab-to-oci is actually quite low. Under the hood, the image relocator already works image per image, it is just not exposed (yet).

**Comment by glyn** [comment]
Ok, that all sounds reasonable, although it might take a while to finalise. May I suggest a practical way forward? cnab-to-oci could provide an equivalent to the `registry` package of github.com/pivotal/image-relocation and then we could simply re-base the implementation of `duffle relocate` on that package. (The `NewLayout` and `Add` methods are not needed to implement `duffle relocate`, but will be useful when `duffle` wants to support creating thick bundles containing OCI image layouts.)
